### PR TITLE
[1.20.5] Network API Cleanup

### LINKED ISSUE #459: isConnected does not convey effectively what it does
The `isConnected` method introduced on the player connection with the networking PR does not convey what it does effectively.

The method currently checks if the connection that is already connected has told the server that it can receive a specific channel/packet `isConnected` does not convey this which makes it hard to search for in the code base I personally spent a bit trying to find such a method and couldn't and had to ask for if there was such a method.

A more effective name could be something along the lines of `canSend`, `canReceive`, `hasChannel`, `hasPacket` all these names have the same thing in common they are more directly related to it being a packet or channel of a packet that is trying to be checked which is what people will be searching for when they want to find it. Could also do something closer to what it used to be with `SimpleChannel`s it could be `isPresent`

### PR DESCRIPTION
**Pull request by Shadows-of-Fire**
This PR is a combination of many distinct network refactors, and are described individually below.
### Motivation
The payload context was previously composed of three (four for config) interfaces, each of which had one implementation per side (a total of 12 classes), on top of that, there were subclasses for each protocol, leading to what was effectively a maze of interfaces.  

Since these interfaces always have one concrete implementation (per-side), and were always constructed at the same time (using the same arguments, the packet listener), it is more reasonable to flatten them into the `IPayloadContext` object, and create two sided implementations of the context, usable by both protocols.

### Changes
The following classes have been removed: 
* `IReplyHandler`
  * `ClientReplyHandler`
  * `ServerReplyHandler`
* `IPacketHandler`
  * `ServerPacketHandler`
  * `ClientPacketHandler`
* `ISynchronizedWorkHandler`
  * `EventLoopSynchronizedWorkHandler`
* `ITaskCompletedHandler`,
* `ConfigurationPayloadContext`
* `IConfigurationPayloadHandler`
* `PlayPayloadContext`
* `IPlayPayloadHandler`

The following methods were added to `IPayloadContext`:
* `IPayloadContext#reply` - replacing `IReplyHandler#send`
* `IPayloadContext#disconnect` - replacing both `IReplyHandler#disconnect` and `IPacketHandler#disconnect`
* `IPayloadContext#handle` - replacing `IPacketHandler#handle` (including overloads)
* `IPayloadContext#enqueueWork` - replacing `ISynchronizedWorkHandler#submitAsync` (including overloads) with implicit exception handling
* `IPayloadContext#finishCurrentTask` - replacing `ITaskCompletedHandler#onTaskCompleted` 
* `IPayloadContext#sender` - replacing `IPayloadContext#player`

All prior uses of `PlayPayloadContext` and `ConfigurationPayloadContext` have been loosened to `IPayloadContext`.

### Benefits

Flattening reduces indirection (consumers of the context can perform their actions immediately with the context object) and deduplicates a sizeable amount of code that was a result of the composition.
Additionally, we can reuse more code instead of fragmenting things on a per-protocol basis.

### Drawbacks
The method `IPayloadContext#finishCurrentTask` is now always visible, instead of only being visible to configuration-protocol payloads.  However, this is acceptable, as the benefit gained from not having to segment the paths is substantial.  We can make a reasonable assumption that users can read (we did so previously, because the method would throw on the client, even in configuration-protocol payloads).

### Motivation
All data within `NetworkRegistry` was static (including the existing non-static fields, due to the instantiation of `NetworkRegistry#INSTANCE`), but accessing this data required going through `NetworkRegistry.getInstance` which was a fairly pointless step.

### Changes
All fields and methods in `NetworkRegistry` were made static.  `NetworkRegistry#INSTANCE` and `NetworkRegistry#getInstance()` were removed.

## Rename of `isConnected` to `hasChannel`
See #459 

### Motivation
The event name is ill-formed for its purpose (which is to register additional configuration task objects to be run during the configuration phase).

### Changes
`OnGameConfigurationEvent` was renamed to `RegisterConfigurationTasksEvent`, which better reflects its purpose. For better compatibility, the event was also loosened to accept `ConfigurationTask` instances instead of only `ICustomConfigurationTask`.

### Motivation
There were many code paths within NetworkRegistry were either exact copies or near-exact copies of one another. By adjusting how we store certain data, we can centralize these code paths and not need multiple copies of the same thing.
### Changes
Code paths that were separated for play and configuration phases have been merged and now access the backing protocol-keyed `PAYLOAD_REGISTRATIONS` registry. Many data structures that also stored objects for both play and configuration phases reflect a similar change.
### Benefits
This allows us to automatically determine the correct map based on the protocol, which is already embedded in the `PacketListener`, instead of needing to check the type of the listener and redirect to the correct map.  
Additionally, if Mojang ever decides to support custom payloads on other protocols, or adds additional protocols which support them, we will not have to perform a major refactor of our data structures.

### CODE REVIEW

**Comment by marchermans** [comment]
I personally don't support this change.
A lot of the removals are personal opinion of the author of this PR on style and architecture (yes my original creation was as well, but the API has now existed for not even one version, you could have added comments to the original 20.4.7-beta PR and request it to be flattened).

There are several usability reductions in this proposal as well (the removal of ICustomConfigurationTask being one of them) we added the indirection and the handling of for example the payload wrapping in neoforge so that a modder only has to interact with his or her payload. Not the packet. The whole core tenant of the API gets now partially shredded. Which is simply not something I want to support.

Simply said, the Networking API is not complete, which has been stated many times over by now. And these interfaces serve a role in the future underlying reworks. Flattening it all only exposes more API surface which was never intended to be public and binds us to structure that we might need to change in the future. Leaving it as separate interfaces hurts nobody but ourselves.

**Comment by Shadows-of-Fire** [comment]
Unfortunately I was mostly unavailable for the duration of the original networking PR, so I could not comment on its implementation at that time.

If you are willing to back up the need for interface composition of `IPayloadContext` we can argue keeping it, but as-is the composition structure is just noise.

I am iffy about my removal of `ICustomConfigurationTask` after updating the existing implementations to match.  I think it may make more sense to retain it as a helper and loosen the event to `ConfigurationTask` (it was previously strongly-typed to `ICustomConfigurationTask`).

File: src/main/java/net/neoforged/neoforge/network/filters/GenericPacketSplitter.java [hunk 1]
```diff
@@ -208,7 +201,7 @@ public enum RemoteCompatibility {
     }
 
     public static RemoteCompatibility getRemoteCompatibility(Connection manager) {
-        return NetworkRegistry.getInstance().isConnected(manager, ConnectionPhase.ANY, SplitPacketPayload.ID) ? RemoteCompatibility.PRESENT : RemoteCompatibility.ABSENT;
+        return NetworkRegistry.hasChannel(manager, null, SplitPacketPayload.ID) ? RemoteCompatibility.PRESENT : RemoteCompatibility.ABSENT;
```
Comment (Shadows-of-Fire) [inline src/main/java/net/neoforged/neoforge/network/filters/GenericPacketSplitter.java:204]: @marchermans was this usage of `ConnectionPhase.ANY` even correct before this change?  I'm thinking it should have been referencing `GenericPacketSplitter#connectionType` instead.

**Comment by Shadows-of-Fire** [comment]
Marking as ready for review.  Description has been updated to reflect the current state of this PR.

**Pull request closed without merging**